Fixed #24229 -- Validate floats representing integers as integers.

jdufresne opened the pull request.

wimfeijen commented:
Wouldn't int(float(str(value))) work as well? In that case we don't need a regex, which may be difficult tot interpret.

jdufresne commented:
> Wouldn't int(float(str(value))) work as well? In that case we don't need a regex, which may be difficult tot interpret.

Actually, that was my original approach. But the implementation causes `models_forms.tests.ModelOtherFieldTests.test_big_integer_field()` to fail. If the input is a very large integer (`9223372036854775808` in the test) going from str -> float -> int can lose precision and potentially change the integer value.

After I saw that this test would fail, I introduced the regex.

wimfeijen commented:
I see. Thanks for pointing this out! Interesting though. 

>>> int(float(str(9223372036854775808)))
9223372036854775808L
>>> int(float(str('9223372036854775808.0')))
9223372036854775808L
>>> int(float(str(9223372036854775808.0)))
9223372036849999872L
>>> str(9223372036854775808.0)
'9.22337203685e+18'

Shall we include such a large number in the test, for clarity?

jdufresne commented:
> Shall we include such a large number in the test, for clarity?

Good idea. I will add test cases for large numbers.

jdufresne commented:
BTW, I'm open to other ideas about implementation if the regex is undesirable.

claudep commented on django/forms/fields.py:
```diff
@@ -259,8 +260,10 @@ def to_python(self, value):
+        value = self.re_decimal.sub('', str(value))
         try:
-            value = int(str(value))
+            value = int(value)
```
What about keeping `str(value)` inside the try block? I think that on Python 2, a unicode input would raise UnicodeEncodeError instead of ValidationError (might be worth to add that in the tests).

jdufresne commented:
> What about keeping str(value) inside the try block? I think that on Python 2, a unicode input would raise UnicodeEncodeError instead of ValidationError (might be worth to add that in the tests).

Indeed. I fixed this and added a unit test to demonstrate the fix. Thanks.

claudep commented:
Thanks, pushed in 3b966c2b73
